### PR TITLE
gdynia data ingestion

### PR DESCRIPTION
**Pull request by Anakin100100**

### CODE REVIEW

**Comment by Anakin100100** [comment]
@wojtodzio I'm a bit at a loss with what is wrong with rswag. I have followed the tutorials to the letter and after generating correct output the UI shows a blank page and throws a bunch of errors because there is no javascript or required images there. Can you look at this? Maybe you can find out what is wrong. One thing that may be worth looking at is the fact that only version that was compatible with rails 7.0.0 is the rswag version from master branch on github and not an official release.

File: node_modules/.yarn-integrity
Comment (wojtodzio) [inline node_modules/.yarn-integrity:1]: I don't think this file should be checked into the repository

File: spec/requests/get_all_institutions_spec.rb
Comment (wojtodzio) [inline spec/requests/get_all_institutions_spec.rb:5]: ✂️

File: spec/requests/get_one_institution_spec.rb
Comment (wojtodzio) [inline spec/requests/get_one_institution_spec.rb:5]: ✂️

File: spec/requests/institutions_spec.rb
Comment (wojtodzio) [inline spec/requests/institutions_spec.rb:4]: It seems like this file uses four space long indents, while all the other files use two spaces

File: spec/services/gdynia_extra_data_ingestor_spec.rb
Comment (wojtodzio) [inline spec/services/gdynia_extra_data_ingestor_spec.rb:7]: It'd be nice to have a test that mocks the data and checks that it's being actually properly processed

File: spec/services/gdynia_extra_data_ingestor_spec.rb
Comment (wojtodzio) [inline spec/services/gdynia_extra_data_ingestor_spec.rb:6]: The convention for describes is just to use the function's name - makes it easier to read multiple tests and 
```suggestion
  describe '#call' do
```

File: yarn.lock
Comment (wojtodzio) [inline yarn.lock:1]: We don't need yarn for api

**Review by wojtodzio — CHANGES_REQUESTED** [review]
Excellent, I genuinely like where this is going. Leaving a few minor comments.
I'm not sure what's wrong with rswag. Unfortunately, I don't have time to look into it right now, we could pair program on it, but I probably won't have time before Monday.

**Comment by Anakin100100** [comment]
@wojtodzio, sure we can try to pair program the rswag on monday. It's not very urgent.

File: yarn.lock
Comment (Anakin100100) [reply yarn.lock:1]: I tried yarn install to maybe install some missing javascript for the rswag but it didn't work

File: app/clients/gdynia_extra_data_client.rb
Comment (wojtodzio) [reply app/clients/gdynia_extra_data_client.rb:33]:
```ruby
stub_request(:get, ".../schools/").to_return(body: "{ ... }")
```

File: db/migrate/20220204083125_add_extra_fields_from_gdynia.rb
Comment (wojtodzio) [inline db/migrate/20220204083125_add_extra_fields_from_gdynia.rb:6]: Look if those shouldn't be ints

File: spec/clients/gdynia_extra_data_client_spec.rb
Comment (wojtodzio) [inline spec/clients/gdynia_extra_data_client_spec.rb:5]:
```suggestion
RSpec.describe 'GdyniaExtraDataClient' do
```

**Review by wojtodzio — DISMISSED** [review]

File: app/clients/gdynia_extra_data_client.rb
Comment (Anakin100100) [reply app/clients/gdynia_extra_data_client.rb]: @wojtodzio when mocking the response only for some of the api request is the WebMock.allow_net_connect! global and has to be later disabled with WebMock.disable_net_connect! or is it scoped and the WebMock.disable_net_connect! is redundant?

**Comment by Anakin100100** [comment]
@wojtodzio @micorix We are ready to merge. @micorix please accept the pull request as your approval is required due to miniscual change in the linting and testing workflow.

File: .github/workflows/ci.yml
Comment (micorix) [inline .github/workflows/ci.yml:12]: Why?

Comment (Anakin100100) [reply .github/workflows/ci.yml:12]: @micorix because we want to run this workflow on every commit. The previous config doesn't run this workflow on pull requests to pull requests and drafts

File: spec/jobs/create_institution_records_job_spec.rb
Comment (wojtodzio) [inline spec/jobs/create_institution_records_job_spec.rb:12]: This can keep net connections enabled if an error occurs between allow and disable. As an alternative, you could add:
```ruby
RSpec.configure do |config|
  config.around(:each, allow_net_connect: true) do |example|
    WebMock.allow_net_connect!
    example.run
  ensure
    WebMock.disable_net_connect!
  end
end
```
to your rails_helper and then add a tag to your test
```ruby
it '...', :allow_net_connect do
  # here net connections will be enabled
end
```

**Review by wojtodzio — APPROVED** [review]
🚀

**Comment by Anakin100100** [comment]
We cannot merge yet. I just discovered that we have a bug that we haven't noticed before. We have only 105 distinct RSPO institution ids among 9937 records which is of course impossible. Most rspo_institution_ids have 100 records associated with them (The payload size per page is 100 institutions in the RSPO api) while a minority has less than 100. This would imply that one record is copied over to an entire batch before inserting to the database. We have to resolve this issue before merging.

**Comment by Anakin100100** [comment]
This is exactly what was happening.

File: app/jobs/create_institution_records_job.rb
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:19]: If there was a bug, when fixing it, you should add a regression test, i.e., a test proving the bug is fixed.

File: app/models/institution.rb
Comment (Anakin100100) [reply app/models/institution.rb:7]: This is not related to the pull request because I discovered the bug when working on data filtering

File: app/jobs/create_institution_records_job.rb
Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:19]: I checked it manually the bug has been fixed. Designing the test for is is going to take me a long time because I'd have to mock the entire RSPO api write the responses and check all fields. I don't have unlimited time to spend on done features. This feature won't be modified when it comes to business logic maybe with regard to fields that are added to the database.

Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:19]: I think that it can be left as is. Doing refactoring to save one or two lines of code is sort of missing the point in my opinion. Using fancy iteration techniqes like each_with_object later leads to bugs like this. Trying to refactor away two lines of perfectly readible code has cost me two hours of debugging later. I would prefer to use simple iteration that is easy to understand and where it is hard to make mistakes like that lead to bugs later.

Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:19]: If this didn't work previously, it means it wasn't properly tested; if it wasn't properly tested, it was never done. It's not about spending time on done features; it's about making sure you're doing high-quality work. 
If you're making real requests from tests, all you have to do is to test how many records this job creates. If you want to stub requests, you don't have to mock the entire RSPO API, you can assume the page's size is 2 and the API contains 3 records. Both things are very easy.

Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:19]: Man, `map` is not fancy :D If you want to write ruby, you need to know `map`, `each_with_object`, `reduce`, and other `Enumerable` methods by heart. There's no other way around it. If you were to use loops, you might as well write this in C. The whole power of Ruby comes from being able to express the very essence of what you're doing in a concise, readable way. Once you get some experience using those higher-order functions, you'll realize it's actually way easier to write and way harder to make mistakes using it - but there's no other way to gain that experience than just to use them.

Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:19]: Look at this this way, your original code was basically this:
```ruby
result = []
i = 0
loop do
  result << transform(data[i])
  break unless data[i]
end
```
The map version is basically this:
```ruby
result = data.map { |obj| transform(obj) }
```
In the first version, you have to create 2 variables, which can lead to mistakes. You have to write a break condition, which can lead to mistakes. You have to mutate an object, which can lead to mistakes. You need to do some thinking to understand the content of the result object at any point in time.
In the second version, there are no intermediate variables, no conditions, no mutations. You immediately see that the result object is an array of the same length as the data array. Sure, you need to understand how `map` works, but this is fundamental in any higher-order language.

Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:19]: The code logic you mention was used when we needed the index to access the hash inside the hash of hashes with stringified index as second key. After they changed it to array of hashes I switched to the logic that is used now.

Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:19]: It worked perfectly fine before the refactoring was ordered. Yesterday during code review we agreed on this being merge ready provided that a few extremely minor things are fixed and they have been. I would prefer to spend time on new features that bring us closer to a goal of it being ready at the end of the week in a tangible way and not on applying refactoring to bugfixes for bugs resulting from refactoring and applying refactoring to refactoring to refactoring.

File: .github/workflows/ci.yml
Comment (micorix) [reply .github/workflows/ci.yml:12]: We can enable CI on more branches if you insist, but I'd rather not do that. Here's why:
1. Does drafts need CI? Besides r u sure draft PRs with master branch as merge target don't trigger workflows?
2. I'm not sure if making PRs to branches with already-open PR to master is such a good thing, coz
  * It may lead to duplicated code reviews 
  * and it encourages contributors to make large PRs
  * not sure if i've seen sth like it before (altho it's not 100% proper argument against it)

Which leads to another thing. I've seen that your PRs contain lots of changes (i.e. this one: +536 −93 LoC). I'm not a backend dev but I think there should be some way to divide huge PRs into lots of smaller ones. That way code review (probably) may be done quicker and development would be more rapid. (Altho we still need to keep in mind assumption that master should be deployable at any time)

Comment (Anakin100100) [reply .github/workflows/ci.yml:12]: 1 Yes you want to see if tests pass from that commit and if linting is ok.
2 Because we have one feature branch at a time the pull request is quite large and some things are done in branches where it doesn't make sense to branch out of master and that smaller thing is finished before the big thing is finished so this smaller thing is merged into the bigger merge. 

You don't have to worry about there being a lot of changes in a single pr because a feature that is being developed is quite robust and contains a lot of autogenerated code related to rswag and a lot of line changes in Gemfile.lock, Gemfile etc. that are ignored during code review. I was doing code review with Wojtek yesterday and it wasn't a problem.

File: app/clients/gdynia_extra_data_client.rb
Comment (wojtodzio) [reply app/clients/gdynia_extra_data_client.rb]: I think it's global, but you'd have to double-check me on that

File: .github/workflows/ci.yml
Comment (wojtodzio) [reply .github/workflows/ci.yml:12]: Yeah, it'd be way better to do smaller PRs and to diverge them from the master, especially when working on independent features - like here, at the very least, swagger could've been a separate PR. It'd make the process way faster, for sure. But what we're doing right now is ok for me for now - I hope once @Anakin100100 gets a little bit more experience with Git, those PRs will start to look better ;)
As for this change, while I do agree we should strive not to make PRs to other PRs, I don't see many disadvantages of not running CI for all branches, and I could see why we would want to have such PRs (e.g., if we ware to use gitflow).

File: app/jobs/create_institution_records_job.rb
Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:19]: I'm not sure I understand. I don't want to block the development, so if you don't know how to use map, we can leave it is as it is and schedule a quick pair programming to fix it next week.

Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:19]: Hmm, I think we misunderstood each other. What I'm asking for is to add a missing test, not to "apply refactoring to bugfixes for bugs resulting from refactoring and applying refactoring to refactoring to refactoring".
It'll probably be enough to change `expect(Institution.where(rspo_institution_type_id: 93).count).to be_positive` to `expect(Institution.where(rspo_institution_type_id: 93).count).to eq(number)`, where `number` is an actual number of created records. This'll probably take you less time than writing a comment.

**Comment by Anakin100100** [comment]
@wojtodzio @micorix We are ready to merge again.

**Review by wojtodzio — APPROVED** [review]
Nice 🚀

**Review by micorix — APPROVED** [review]

**Pull request merged**
